### PR TITLE
basic support for try except

### PR DESCRIPTION
**Pull request by sklam**
Implements https://github.com/numba/numba/issues/4898#issuecomment-559110992

### CODE REVIEW

**Comment by sklam** [comment]
I'll add user docs in a different PR.

**Comment by seibert** [comment]
running through internal CI (`numba_58`)

**Comment by seibert** [comment]
CI is failing for Python 3.7 and 3.8 with one failing test:
```
[2019-12-09 10:30:34,437] {bash_operator.py:137} INFO - ======================================================================
[2019-12-09 10:30:34,437] {bash_operator.py:137} INFO - FAIL: test_unsupported_op_code (numba.tests.test_dataflow.TestDataFlow)
[2019-12-09 10:30:34,437] {bash_operator.py:137} INFO - ----------------------------------------------------------------------
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO - Traceback (most recent call last):
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\tests\test_dataflow.py", line 206, in test_unsupported_op_code
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     compile_isolated(pyfunc, (), flags=flags)
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 180, in compile_isolated
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     flags, locals)
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 541, in compile_extra
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     return pipeline.compile_extra(func)
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 328, in compile_extra
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     return self._compile_bytecode()
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 387, in _compile_bytecode
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     return self._compile_core()
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 367, in _compile_core
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     raise e
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler.py", line 358, in _compile_core
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     pm.run(self.state)
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler_machinery.py", line 331, in run
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     raise patched_exception
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler_machinery.py", line 322, in run
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     self._runPass(idx, pass_inst, state)
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler_lock.py", line 32, in _acquire_compile_lock
[2019-12-09 10:30:34,438] {bash_operator.py:137} INFO -     return func(*args, **kwargs)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler_machinery.py", line 281, in _runPass
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     mutated |= check(pss.run_pass, internal_state)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\compiler_machinery.py", line 268, in check
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     mangled = func(compiler_state)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\untyped_passes.py", line 113, in run_pass
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     post_proc.run()
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\postproc.py", line 78, in run
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     self.func_ir.blocks = transforms.canonicalize_cfg(self.func_ir.blocks)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\transforms.py", line 303, in canonicalize_cfg
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     return canonicalize_cfg_single_backedge(blocks)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\transforms.py", line 232, in canonicalize_cfg_single_backedge
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     cfg = compute_cfg_from_blocks(blocks)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\analysis.py", line 243, in compute_cfg_from_blocks
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     cfg.add_edge(k, target)
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -   File "C:\Users\ci\64\Miniconda3\envs\testenv_58951d2a-c31b-4813-9839-20eec3295a52\lib\site-packages\numba\controlflow.py", line 86, in add_edge
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO -     assert dest in self._nodes
[2019-12-09 10:30:34,439] {bash_operator.py:137} INFO - AssertionError: Failed in object mode pipeline (step: processing IR)
```

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/4902#issuecomment-563327457 is because of empty try-except:

```python
try:
    pass
except:
    pass
```

Fixing it now

File: numba/tests/test_try_except.py [hunk 1]
```diff
@@ -208,6 +208,165 @@ def udt(x):
...
+                    return 12
+            if x == 1:
+                raise ValueError("exception")
+
+        with captured_stdout() as stdout:
+            res = foo(10)
+
+        self.assertEqual(stdout.getvalue().split(), ["CAUGHT",],)
+
+    # this is broken
```
Comment (esc) [inline numba/tests/test_try_except.py]: This comment seems useless, makes me ask the question, what exactly is broken. And maybe it isn't broken anymore.

File: numba/tests/test_try_except.py [hunk 2]
```diff
@@ -208,6 +208,165 @@ def udt(x):
...
+        @njit
+        def bar(z):
+            return next(foo(z))
+
+        with captured_stdout() as stdout:
+            res = bar(10)
+
+        self.assertEqual(stdout.getvalue().split(), ["CAUGHT",],)
+
+    # segfault, uncaught runaway recursion?
```
Comment (esc) [inline numba/tests/test_try_except.py]: Maybe fine by now?

File: numba/tests/test_try_except.py (hunk 1 above)
Comment (esc) [reply numba/tests/test_try_except.py]: Oups, wrong window, nevermind.

Comment (sklam) [reply numba/tests/test_try_except.py]: oops, that's left over comments from tests written by @stuartarchibald. I thought I removed them all.

File: numba/tests/test_try_except.py (hunk 2 above)
Comment (esc) [reply numba/tests/test_try_except.py]: Wrong window, nevermind.

**Comment by sklam** [comment]
Added checks to disallow storing the exception object into a user variable; i.e. `Exception as e`

**Comment by sklam** [comment]
I will be adding docs tmr

**Comment by stuartarchibald** [comment]
@sklam these could probably do with looking at https://github.com/stuartarchibald/numba/commit/c75ec26aa8f6a44b2dd770bc53c57c8723574886#diff-00b0797d8859391b8f407feea491d0b4R362-R494, one causes the CFG to break, the others are `RuntimeError`s that shouldn't be. Thanks.

File: numba/byteflow.py [hunk 3]
```diff
@@ -15,6 +16,43 @@
 _logger = logging.getLogger(__name__)
 
 
+_EXCEPT_STACK_OFFSET = 6
+_FINALLY_POP = _EXCEPT_STACK_OFFSET if PYVERSION >= (3, 8) else 1
+
+
+@total_ordering
+class BlockKind(object):
```
Comment (stuartarchibald) [inline numba/byteflow.py]: Was enum not suitable?

File: numba/byteflow.py [hunk 4]
```diff
@@ -67,7 +105,23 @@ def run(self):
                     # Terminated?
                     if state.has_terminated():
                         break
-                    # Not yet?
+                    elif state.has_active_try():
+                        # Is in an *try* block
```
Comment (stuartarchibald) [inline numba/byteflow.py]:
```suggestion
                        # Is in a *try* block
```

File: numba/byteflow.py [hunk 5]
```diff
@@ -974,6 +1095,29 @@ def outgoing_phis(self):
         """
         return self._outgoing_phis
 
+    @property
+    def blockstack_initial(self):
+        """A copy of the initial state of the blockstack
```
Comment (stuartarchibald) [inline numba/byteflow.py]: Is this a copy or a reference?

File: numba/byteflow.py [hunk 6]
```diff
@@ -1080,8 +1276,20 @@ def get_outgoing_states(self):
                 self._outgoing_phis[phi] = edge.stack[i]
         return ret
 
+    def get_outgoing_edgepushed(self):
+        """
+        Returns
+        -------
+        Dict[int, int]
+            where keys are the PC
+                  values are the edge-pushed stack values
```
Comment (stuartarchibald) [inline numba/byteflow.py]:
```suggestion
            values are the edge-pushed stack values
```

File: numba/interpreter.py [hunk 7]
```diff
@@ -127,12 +131,37 @@ def interpret(self, bytecode):
...
+                    k = v.name
+                    if k in excvars:
+                        excvars.add(varname)
+        # Filter out the user variables.
+        uservar = list(filter(lambda x: not x.startswith('$'), excvars))
+        if uservar:
+            # Complain about the first user-variable storing an exception
+            first = uservar[0]
+            loc = self.current_scope.get(first).loc
+            msg = "Exception object cannot be stored into variable."
```
Comment (stuartarchibald) [inline numba/interpreter.py]: Can the variable name become part of the string?

File: numba/interpreter.py [hunk 8]
```diff
@@ -171,9 +216,71 @@ def _start_new_block(self, offset):
...
+    def _insert_try_block_end(self):
+        """Insert IR-nodes to mark the end of a `try` block.
+        """
+        gv_fn = ir.Global(
+            "end_try_block", eh.end_try_block, loc=self.loc,
+        )
+        end_try_name = 'end_try_fn'
+        self.store(value=gv_fn, name=end_try_name, redefine=True)
+        end_try = ir.Expr.call(self.get(end_try_name), (), (), loc=self.loc)
+        self.store(value=end_try, name='$end_try', redefine=True)
```
Comment (stuartarchibald) [inline numba/interpreter.py]: not essential but these are sufficiently similar they could share most of the lines.

File: numba/interpreter.py [hunk 9]
```diff
@@ -171,9 +216,71 @@ def _start_new_block(self, offset):
...
+        edgepushed = self.dfainfo.outgoing_edgepushed.get(endblk)
+        # Note: the last value on the stack is the exception value
+        # Note: due to the current limitation, all exception variables are None
+        if edgepushed:
+            scope = self.current_scope
+            const_none = ir.Const(value=None, loc=self.loc)
+            # For each variable going to the handler block.
+            for var in edgepushed:
+                if var in self.definitions:
+                    raise AssertionError(
```
Comment (stuartarchibald) [inline numba/interpreter.py]: How is this path reached? Is it tested?

File: numba/interpreter.py [hunk 10]
```diff
@@ -171,9 +216,71 @@ def _start_new_block(self, offset):
...
+                        "exception variable CANNOT be defined by other code",
+                    )
+                self.store(value=const_none, name=var)
+                self._exception_vars.add(var)
+
+    def _insert_exception_check(self):
+        """Called before the end of a block to inject checks if
```
Comment (stuartarchibald) [inline numba/interpreter.py]: `checks if....` what?!

File: numba/interpreter.py [hunk 11]
```diff
@@ -1041,7 +1186,16 @@ def op_MAKE_FUNCTION(self, inst, name, code, closure, annotations, kwdefaults, d
...
             else:
                 defaults = self.get(defaults)
-        fcode = self.definitions[code][0].value
+
+        assume_code_const = self.definitions[code][0]
+        if not isinstance(assume_code_const, ir.Const):
+            msg = (
+                "Unsupported use of closure. "
+                "Probably caused by complex control-flow constructs; "
+                "i.e. try-except"
```
Comment (stuartarchibald) [inline numba/interpreter.py]:
```suggestion
                "e.g. try-except"
```

File: numba/targets/callconv.py [hunk 12]
```diff
@@ -376,12 +379,63 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
...
+        builder.store(new, try_state_ptr)
+
+    def unset_try_status(self, builder):
+        try_state_ptr = self._get_try_state(builder)
+        # Decrement try depth
+        old = builder.load(try_state_ptr)
+        new = builder.sub(old, old.type(1))
+        builder.store(new, try_state_ptr)
+
+        # XXX: reset exception state
```
Comment (stuartarchibald) [inline numba/targets/callconv.py]: needed? TODO?

**Comment by stuartarchibald** [comment]
This:
```python
def bar(x):
    return np.asarray(list(reversed(x.tolist())))

@njit
def test_objmode():
    x = np.arange(5)
    y = np.zeros_like(x)
    try:
        with objmode(y='intp[:]'):  # annotate return type
            # this region is executed by object-mode.
            y += bar(x)
    except:
        pass
    return y
```
gives:
```
numba/numba/interpreter.py", line 837, in op_CALL_FUNCTION_KW
    nkeys = len(keys)
UnboundLocalError: Failed in nopython mode pipeline (step: analyzing bytecode)
local variable 'keys' referenced before assignment
```

**Comment by stuartarchibald** [comment]
This:
```python
@njit
def udt():
    x = 1
    y = 0
    try:
        x / y
    except Exception:
        pass

udt()
```
raises `ZeroDivisionError` when it shouldn't?

File: numba/byteflow.py (hunk 3 above)
Comment (sklam) [reply numba/byteflow.py]: I needed stricter typing at comparison to prune out bugs. Tried enum but it was not strict enough.

File: numba/byteflow.py (hunk 5 above)
Comment (sklam) [reply numba/byteflow.py]: A copy because it was copied at the constructor.

File: numba/interpreter.py (hunk 9 above)
Comment (sklam) [reply numba/interpreter.py]: It's me saying it should be unreachable. But I am not sure.

File: numba/targets/callconv.py (hunk 12 above)
Comment (sklam) [reply numba/targets/callconv.py]: Needed. Updated to a note.

File: numba/byteflow.py [hunk 13]
```diff
@@ -587,8 +587,14 @@ def op_YIELD_VALUE(self, state, inst):
         state.push(res)
 
     def op_RAISE_VARARGS(self, state, inst):
+        in_exc_block = state.get_top_block("EXCEPT") is not None
         if inst.arg == 0:
             exc = None
+            if in_exc_block:
+                raise UnsupportedError(
+                    "Re-raise exception not supported, yet.",
```
Comment (stuartarchibald) [inline numba/byteflow.py]: How about "The re-raising of an exception is not yet supported" ?

Comment (sklam) [reply numba/byteflow.py]: applied

**Comment by sklam** [comment]
* https://github.com/numba/numba/pull/4902/commits/f127335b25a66147fb643992aacde6b6952cc24f fixes https://github.com/numba/numba/pull/4902#issuecomment-564569632 and it's not an elegant way but will do the job before all the old-style impl is updated
* https://github.com/numba/numba/pull/4902/commits/6af3ef1c727d1e5a09b43fbf823ddd023aa40d74 fixes https://github.com/numba/numba/pull/4902#issuecomment-564569632

**Comment by seibert** [comment]
now has conflict with master

**Comment by esc** [comment]
🎉

**Comment by mobassir94** [comment]
i tried the code above : 

@njit
def udt():
    x = 1
    y = 0
    try:
        x / y
    except Exception:
        pass

udt()

in numba version 0.47 and 0.48 and it still throws me error :  "UnsupportedError: Failed in nopython mode pipeline (step: analyzing bytecode)
'try' block not supported until python3.7 or later"

how to solve this issue?

**Comment by esc** [comment]
@mobassir94 I just tried the snippet you suggested on the current Numba master `0.50.0dev0-7-g917f582433-dirty` and it seems to be fine. Could you give it a try?

**Comment by mobassir94** [comment]
@esc which one?
i tried this : 

`@njit
def udt():
    x = 1
    y = 0
    try:
        x / y
    except Exception:
        pass

udt()`

and it didn't work at all in kaggle kernel

another code that i was trying is this one : 

`%%time
from langdetect import detect
count = 0
noise = []
non_eng = []
for i in range(len(df_train_full)):
    try:
        lang = detect(df_train_full.comment_text[i])
        if(lang != 'en'):
            count +=1
            non_eng.append(i)
    except:
        noise.append(i)
`

and the error comes for this line of code : lang = detect(df_train_full.comment_text[i])
i can't use this library : https://pypi.org/project/langdetect/ in try  block,will you please give it a try and see if you can make it work?
the dataset i was using is this one : https://www.kaggle.com/c/jigsaw-multilingual-toxic-comment-classification/data

**Comment by esc** [comment]
@mobassir94 thanks for following up. We appreciate your input in this matter. I am not sure what you mean "it didn't work at all in kaggle kernel"?! I just tried it again locally:

```
💥 zsh» cat foo033.py
from numba import njit


@njit
def udt():
    x = 1
    y = 0
    try:
        x / y
    except Exception:
        pass


udt()
💥 zsh» python foo033.py
💥 zsh» echo $?
0
💥 zsh» git desc
0.50.0dev0-7-g917f582433
```

 As you can see I am not receiving any error message. Perhaps you could submit the full backtrace of the exception? Also, what Python version are you using? As can be seen on https://pypi.org/project/numba/ Numba only supports Pythonn 3.6 and later.

As for the issues with `langdetect`, only a subset of NumPy and Python functions/features are supported by Numba, documentation on those supported can be found here.
 * [NumPy Features](http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html)
 * [Python Features](http://numba.pydata.org/numba-doc/latest/reference/pysupported.html)

Effectively this means you can not use `langdetect` in functions that you would like to compile with  Numba.

**Comment by gmarkall** [comment]
@mobassir94 Could you please open an issue for further discussion please? It will also be helpful for you to report the numba version used in the Kaggle kernel, i.e. `numba.__version__` - many thanks!

**Comment by mobassir94** [comment]
dear @gmarkall  as from last comment of @esc  it is clear to me that numba doesn't support outside package like langdetect,i was using langdetect in try block so it was not working,i guess the program i tried to execute on gpu can't be done using numba,thanks for your comment

**Comment by segatrade** [comment]
Hello!
I run:
```
try:
    xx_inv = np.linalg.inv(xx)
except np.linalg.LinAlgError:
    return [np.nan], [[np.nan, np.nan]]
```
It gives:
```
numba.core.errors.UnsupportedError: Failed in nopython mode pipeline (step: analyzing bytecode)
'try' block not supported until python3.7 or later
```
I have python3.6, but in my case this is requirement. Is any way to use try on py3.6 with numba or replacement for try?

**Pull request merged**